File: Projects/codechef/BSTOPS.java
import java.io.*;
import java.util.*;

class BSTOPS {
	private static class Node {
		Node lt, rt;
		int x;
		Node(int x) { this.x = x; }
	}
	static Node insert(Node root, int x, long p) {
		if (root == null) {
			System.out.println(p);
			return new Node(x);
		} else if (x < root.x)
			root.lt = insert(root.lt, x, p * 2);
		else if (x > root.x)
			root.rt = insert(root.rt, x, p * 2 + 1);
		return root;
	}
	static Node delete(Node root, int x, long p) {
		if (x < root.x)
			root.lt = delete(root.lt, x, p * 2);
		else if (x > root.x)
			root.rt = delete(root.rt, x, p * 2 + 1);
		else {
			if (p > 0)
				System.out.println(p);
			if (root.lt == null)
				return root.rt;
			else if (root.rt == null)
				return root.lt;
			else {
				Node successor = root.rt;
				while (successor.lt != null)
					successor = successor.lt;
				root.x = successor.x;
				root.rt = delete(root.rt, successor.x, -1);
			}
		}
		return root;
	}
	public static void main (String[] args) throws IOException {
		Scanner in = new Scanner(System.in);
		int n = in.nextInt();
		Node root = new Node(Integer.MIN_VALUE);
		while (n-- > 0) {
			char c = in.next().charAt(0);
			int x = in.nextInt();
			switch (c) {
			case 'i':
				insert(root, x, 0);
				break;
			case 'd':
				delete(root, x, 0);
				break;
			}
		}
	}
}

File: Projects/codechef/BSTOPS.kt
import java.io.*
import java.util.*

internal object BSTOPS {
    private class Node internal constructor(internal var x: Int, internal var p: Long) {
        internal var lt: Node? = null
        internal var rt: Node? = null
    }

    private fun insert(root: Node?, x: Int, p: Long): Node {
        if (root == null) {
            System.out.println(p)
            return Node(x, p)
        } else if (x < root.x)
            root.lt = insert(root.lt, x, p * 2)
        else if (x > root.x)
            root.rt = insert(root.rt, x, p * 2 + 1)
        return root
    }

    private fun delete(root: Node, x: Int): Node? {
        if (x < root.x)
            root.lt = delete(root.lt!!, x)
        else if (x > root.x)
            root.rt = delete(root.rt!!, x)
        else {
            System.out.println(root.p)
            if (root.lt == null)
                return root.rt
            else if (root.rt == null)
                return root.lt
            else {
                var successor = root.rt
                while (successor!!.lt != null)
                    successor = successor.lt
                root.x = successor.x
                root.rt = delete(root.rt!!, successor.x)
            }
        }
        return root
    }

    @Throws(IOException::class)
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        var n = stdin.nextInt()
        var root: Node? = null
        while (n-- > 0) {
            val c = stdin.next().charAt(0)
            val x = stdin.nextInt()
            when (c) {
                'i' -> root = insert(root, x, 1)
                'd' -> root = delete(root!!, x)
            }
        }
    }
}
